src/core/thread/mesh_forwarder_ftd.cpp

```diff
@@ -402,6 +402,101 @@ Error MeshForwarder::UpdateMeshRoute(Message &aMessage)
     return error;
 }
 
+Error MeshForwarder::ContextIdRouteLookup(uint8_t aContextId, AnycastType aType, uint16_t &aMeshDest) const
+{
+    Lowpan::Context                 context;
+    NetworkData::Iterator           iterator = NetworkData::kIteratorInit;
+    NetworkData::OnMeshPrefixConfig config;
+    uint8_t                         bestCost = Mle::kMaxRouteCost;
+    uint16_t                        bestDest = Mac::kShortAddrInvalid;
+    uint8_t                         routerId;
+
+    SuccessOrExit(Get<NetworkData::Leader>().GetContext(aContextId, context));
+
+    while (Get<NetworkData::Leader>().GetNextOnMeshPrefix(iterator, config) == kErrorNone)
+    {
+        uint16_t        server16 = config.mRloc16;
+        uint8_t         curCost  = 0x00;
+        const Neighbor *neighbor;
+
+        if (config.GetPrefix() != context.mPrefix)
+        {
+            continue;
+        }
+
+        switch (aType)
+        {
+        case kAnycastDhcp6Agent:
+            if (!(config.mDhcp || config.mConfigure))
+            {
+                continue;
+            }
+            break;
+        case kAnycastNeighborDiscoveryAgent:
+            if (!config.mNdDns)
+            {
+                continue;
+            }
+            break;
+        }
+
+        // Path cost
+        curCost = Get<Mle::MleRouter>().GetCost(server16);
+
+        if (!Mle::MleRouter::IsActiveRouter(server16))
+        {
+            // Assume best link between remote child server and its parent.
+            curCost += 1;
+        }
+
+        // Cost if the server is direct neighbor.
+        neighbor = Get<NeighborTable>().FindNeighbor(server16);
+
+        if (neighbor != nullptr && neighbor->IsStateValid())
+        {
+            uint8_t cost;
+
+            if (!Mle::MleRouter::IsActiveRouter(server16))
+            {
+                // Cost calculated only from Link Quality In as the parent only maintains
+                // one-direction link info.
+                cost = Mle::MleRouter::LinkQualityToCost(neighbor->GetLinkInfo().GetLinkQuality());
+            }
+            else
+            {
+                cost = Get<Mle::MleRouter>().GetLinkCost(Mle::Mle::RouterIdFromRloc16(server16));
+            }
+
+            // Choose the minimum cost
+            if (cost < curCost)
+            {
+                curCost = cost;
+            }
+        }
+
+        if ((bestDest == Mac::kShortAddrInvalid) || (curCost < bestCost))
+        {
+            bestDest = server16;
+            bestCost = curCost;
+        }
```

Comment:
This code is extremely similar to the route cost calculation code in `MeshForwarder::GetDestinationRlocByServiceAloc`. 
Should we add a method for this common logic?

Reply:
Updated.